### PR TITLE
Add TF whisper

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

TF Port of the pytorch whisper implementation

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: tests/test_modeling_tf_common.py [hunk 1]
```diff
@@ -1223,8 +1240,17 @@ def test_save_load_after_resize_token_embeddings(self):
 
             # fetch the output for an input exclusively made of new members of the vocabulary
             inputs_dict = copy.deepcopy(original_inputs_dict)
-            new_vocab_input_ids = ids_tensor(inputs_dict["input_ids"].shape, new_tokens_size)
```
Comment (amyeroberts) [inline tests/test_modeling_tf_common.py:1226]: Slight reworking below as previously was trying to get an item from a key that doesn't always exist (input_ids)

File: src/transformers/generation_tf_utils.py [hunk 2]
```diff
@@ -1287,6 +1332,28 @@ def adjust_logits_during_generation(
         elif cur_len == max_length - 1 and forced_eos_token_id is not None:
             vocab_range = tf.constant(range(vocab_size))
             return tf.where(vocab_range != forced_eos_token_id, -1e8, logits)
+
+        elif suppress_tokens is not None:
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:1336]: I actually don't think we need those statements here in the code as the `adjust_logits_during_generation` is just a hack we've implemented for Marian if I recall correctly. So I think just having them implemented as logits processors should be good enough. cc @gante

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 3]
```diff
@@ -0,0 +1,1352 @@
...
+        # If the user passed a tuple for encoder_outputs, we wrap it in a BaseModelOutput when return_dict=True
+        elif return_dict and not isinstance(encoder_outputs, BaseModelOutput):
+            encoder_outputs = BaseModelOutput(
+                last_hidden_state=encoder_outputs[0],
+                hidden_states=encoder_outputs[1] if len(encoder_outputs) > 1 else None,
+                attentions=encoder_outputs[2] if len(encoder_outputs) > 2 else None,
+            )
+
+        # downsample encoder attention mask
+        # decoder_attention_mask = None
```
Comment (patrickvonplaten) [inline src/transformers/models/whisper/modeling_tf_whisper.py:1042]: maybe remove comment here?

File: tests/generation/test_generation_tf_logits_process.py [hunk 4]
```diff
@@ -331,6 +334,86 @@ def test_forced_eos_token_logits_processor(self, use_xla):
...
+        # Check that suppress_tokens are suppressed and others are not
+        cur_len = 5
+        input_ids = ids_tensor((batch_size, cur_len), vocab_size=20)
+        scores = self._get_uniform_logits(batch_size, vocab_size)
+        scores = logits_processor(input_ids, scores, cur_len)
+        self.assertTrue(tf.math.reduce_all(tf.math.is_inf(tf.gather(scores, suppress_tokens, axis=1))))
+        self.assertFalse(tf.math.reduce_any(tf.math.is_inf(tf.gather(scores, keep_tokens, axis=1))))
+
+    @parameterized.expand([(False,), (True,)])
+    def test_force_tokens_logits_processor(self, use_xla):
```
Comment (patrickvonplaten) [inline tests/generation/test_generation_tf_logits_process.py:386]: very cool that they work with XLA!

File: tests/models/whisper/test_modeling_tf_whisper.py [hunk 5]
```diff
@@ -0,0 +1,865 @@
...
+            input_features,
+            do_sample=False,
+        )
+        transcript = processor.batch_decode(generated_ids, skip_special_tokens=True)[0]
+
+        EXPECTED_TRANSCRIPT = " Mr. Quilter is the apostle of the middle classes and we are glad"
+        self.assertEqual(transcript, EXPECTED_TRANSCRIPT)
+
+    @slow
+    def test_large_generation_multilingual(self):
```
Comment (patrickvonplaten) [inline tests/models/whisper/test_modeling_tf_whisper.py:833]: Could we also add 1 or 2 tests for batched generation? E.g. copy-pasting those: https://github.com/huggingface/transformers/blob/45e14038f20d7f04574af4ce8356bab11e3d6741/tests/models/whisper/test_modeling_whisper.py#L1005

File: tests/models/whisper/test_modeling_tf_whisper.py [hunk 6]
```diff
@@ -0,0 +1,865 @@
...
+                1.1624, 0.6194, 1.0757, 1.8259, 2.4076, 1.6601, 2.3503, 1.3376,
+                1.9891, 1.8635, 3.8931, 5.3699, 4.4772, 3.9184
+            ]
+        )
+        # fmt: on
+
+        self.assertTrue(np.allclose(logits[0, 0, :30], EXPECTED_LOGITS, atol=1e-4))
+
+    @slow
+    def test_tiny_en_generation(self):
```
Comment (patrickvonplaten) [inline tests/models/whisper/test_modeling_tf_whisper.py:777]: Also it should work with XLA compiled generation no?

File: src/transformers/models/rag/modeling_tf_rag.py [hunk 7]
```diff
@@ -1256,6 +1256,8 @@ def extend_enc_output(tensor, num_beams=None):
                 eos_token_id=eos_token_id,
                 forced_bos_token_id=None,
                 forced_eos_token_id=None,
+                supress_tokens=None,  # Not used for RAG, were added after the whisper pull request
+                begin_supress_tokens=None,  # Not used for RAG, were added after the whisper pull request
```
Comment (ArthurZucker) [inline src/transformers/models/rag/modeling_tf_rag.py:1260]: I think this is not needed anymore as the default were set to `None` in the generate function

File: tests/generation/test_generation_tf_logits_process.py [hunk 8]
```diff
@@ -331,6 +334,86 @@ def test_forced_eos_token_logits_processor(self, use_xla):
...
+        if use_xla:
+            logits_processor = tf.function(logits_processor, jit_compile=True)
+
+        # Check that suppress_tokens are suppressed and others are not
+        cur_len = 5
+        input_ids = ids_tensor((batch_size, cur_len), vocab_size=20)
+        scores = self._get_uniform_logits(batch_size, vocab_size)
+        scores = logits_processor(input_ids, scores, cur_len)
+        self.assertTrue(tf.math.reduce_all(tf.math.is_inf(tf.gather(scores, suppress_tokens, axis=1))))
+        self.assertFalse(tf.math.reduce_any(tf.math.is_inf(tf.gather(scores, keep_tokens, axis=1))))
```
Comment (ArthurZucker) [inline tests/generation/test_generation_tf_logits_process.py:383]: Very nice tests, the pytorch version doesn't have these! I should open a PR to add that! 👍🏻

File: tests/models/whisper/test_modeling_tf_whisper.py (hunk 6 above)
Comment (amyeroberts) [reply tests/models/whisper/test_modeling_tf_whisper.py:777]: Not yet, unfortunately. I'm going to add a test for XLA generation that we can skip until it's working
